### PR TITLE
[eas-cli] add build:cancel command

### PR DESCRIPTION
**Pull request by wkozyra95**
# Why

Cancel builds from cli

# How

Add `eas build:cancel` command:
- takes build Id as an argument
- if an argument is not provided it's displing select prompt with a list of unfinished builds (or confirm prompt if it's only one build)

```
> eas build:cancel --help
cancel an unfinished build

USAGE
  $ eas build:cancel [BUILD_ID]
```

# Test Plan

Run builds and cancel them

2021-01-25-133653_1354x474_scrot
2021-01-25-134306_1210x124_scrot

### CODE REVIEW

File: packages/eas-cli/src/commands/build/cancel.ts [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        value: build.id,
+      }))
+    );
+    log(buildId);
+    return buildId;
+  }
+}
+
+export default class BuildCancel extends Command {
+  static description = 'cancel an unfinished build';
```
Comment (dsokal) [inline packages/eas-cli/src/commands/build/cancel.ts:97]:
```suggestion
  static description = 'Cancel a build.';
```

File: packages/eas-cli/src/commands/build/cancel.ts [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+  static args = [{ name: 'BUILD_ID' }];
+
+  async run() {
+    const { buildId: buildIdFromArg } = this.parse(BuildCancel).args;
+
+    const projectDir = (await findProjectRootAsync()) ?? process.cwd();
+    const projectId = await getProjectIdAsync(projectDir);
+    const fullProjectName = await getProjectFullNameAsync(projectDir);
```
Comment (dsokal) [inline packages/eas-cli/src/commands/build/cancel.ts:106]:
```suggestion
    const projectFullName = await getProjectFullNameAsync(projectDir);
```

File: packages/eas-cli/src/commands/build/cancel.ts [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+  const startTime = new Date(build.createdAt).toLocaleString();
+  const status = chalk.blue(build.status === BuildStatus.InQueue ? 'in-queue' : 'in-progress');
+  return `${platform} Started at: ${startTime}, Status: ${status}, Id: ${build.id}`;
+}
+
+async function selectBuildToCancelAsync(
+  projectId: string,
+  fullProjectName: string
+): Promise<string | null> {
+  const spinner = ora().start('Fetching the unfinished builds…');
```
Comment (dsokal) [inline packages/eas-cli/src/commands/build/cancel.ts:58]:
```suggestion
  const spinner = ora().start('Fetching uncompleted builds…');
```

File: packages/eas-cli/src/commands/build/cancel.ts [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    spinner.stop();
+    builds = [...inQueueBuilds, ...inProgressBuilds];
+  } catch (error) {
+    spinner.fail(
+      `Something went wrong and we couldn't fetch the builds for the project ${fullProjectName}.`
+    );
+    throw error;
+  }
+  if (builds.length === 0) {
+    log.warn(`There are no unfinished builds for the project ${fullProjectName}.`);
```
Comment (dsokal) [inline packages/eas-cli/src/commands/build/cancel.ts:74]: I'd prefer to move this log outside of this function (to `if (!buildId) { ... }`).
`BuildCancel.run` is the main function and code like `if (!buildId} { return; }` doesn't say anything. I had to check the implementation of `selectBuildToCancelAsync` to see what's the case if there aren't any uncompleted builds.

Comment (dsokal) [reply packages/eas-cli/src/commands/build/cancel.ts:74]:
```suggestion
    log.warn(`There aren't any uncompleted builds for project ${fullProjectName}.`);
```

File: packages/eas-cli/src/commands/build/cancel.ts (hunk 3 above)
Comment (dsokal) [reply packages/eas-cli/src/commands/build/cancel.ts:58]: I'm not an expert but maybe `uncompleted` sounds better.

File: packages/eas-cli/src/commands/build/cancel.ts [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+    const projectDir = (await findProjectRootAsync()) ?? process.cwd();
+    const projectId = await getProjectIdAsync(projectDir);
+    const fullProjectName = await getProjectFullNameAsync(projectDir);
+
+    const buildId = buildIdFromArg || (await selectBuildToCancelAsync(projectId, fullProjectName));
+    if (!buildId) {
+      return;
+    }
+
```
Comment (dsokal) [inline packages/eas-cli/src/commands/build/cancel.ts:112]: What if the user passed the build id as an argument but it's not valid / it doesn't exist in the database?

File: packages/eas-cli/src/commands/build/cancel.ts [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    const spinner = ora().start('Canceling the build…');
+    try {
+      const { status } = await cancelBuildAsync(buildId);
+      if (status === BuildStatus.Canceled) {
+        spinner.succeed('Build canceled');
+      } else {
+        spinner.text = 'Build already finished';
```
Comment (dsokal) [inline packages/eas-cli/src/commands/build/cancel.ts:119]:
```suggestion
        spinner.text = 'Build is already completed';
```

Comment (dsokal) [reply packages/eas-cli/src/commands/build/cancel.ts:119]: Maybe print the build details in this case?

File: packages/eas-cli/src/commands/build/cancel.ts [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+        spinner.stopAndPersist();
+      }
+    } catch (error) {
+      spinner.fail(
+        `Something went wrong and we couldn't fetch the last build for the project ${fullProjectName}`
```
Comment (dsokal) [inline packages/eas-cli/src/commands/build/cancel.ts:124]:
```suggestion
        `Something went wrong and we couldn't cancel your build ${buildId}`
```

File: packages/eas-cli/src/commands/build/cancel.ts [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+          }
+        `,
+        { buildId }
+      )
+      .toPromise()
+  );
+  return data.build!.cancel;
+}
+
+const appPlatformMap = {
```
Comment (dsokal) [inline packages/eas-cli/src/commands/build/cancel.ts:40]: Move this to `build/constants.ts`

File: packages/eas-cli/src/commands/build/cancel.ts [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+  [AppPlatform.Android]: Platform.Android,
+  [AppPlatform.Ios]: Platform.iOS,
+};
+
+function formatUnfinishedBuild(
```
Comment (dsokal) [inline packages/eas-cli/src/commands/build/cancel.ts:45]: I know that this function is only used in this file but maybe `build/utils/printBuildInfo.ts` would be a better place for it? That module contains all the functions that can print build details.

**Review by dsokal — CHANGES_REQUESTED** [review]

File: packages/eas-cli/src/commands/build/cancel.ts (hunk 8 above)
Comment (wkozyra95) [reply packages/eas-cli/src/commands/build/cancel.ts:40]: I prefer to keep it here as long we have duplicate types for apiv2 and graphql api

File: packages/eas-cli/src/commands/build/cancel.ts (hunk 9 above)
Comment (wkozyra95) [reply packages/eas-cli/src/commands/build/cancel.ts:45]: the same as appPlatformMap

File: packages/eas-cli/src/commands/build/cancel.ts (hunk 5 above)
Comment (wkozyra95) [reply packages/eas-cli/src/commands/build/cancel.ts:124]: sql error was printed in that case, I added fetch above and better error message

**Review by dsokal — APPROVED** [review]

**Pull request merged**